### PR TITLE
feat: enable FFT for xHilbertTransform to enhance speed for arrays with lengths that are powers of 2

### PR DESCRIPTION
**Pull request by josoriom**

### CODE REVIEW

File: src/x/xHilbertTransform.ts [hunk 1]
```diff
@@ -1,28 +1,32 @@
...
+  if (!Array.isArray(x)) {
+    throw new Error('x must be a vector');
+  }
+  const n = x.length;
+  const fft = new FFT(n);
+  const v = [1, ...Array(n / 2 - 1).fill(2), 1, ...Array(n / 2 - 1).fill(0)];
+  const X = new Float64Array(n * 2);
+  fft.realTransform(X, x);
+  for (let i = 0; i < n; i++) {
+    X[i * 2] = X[i * 2] * v[i];
```
Comment (lpatiny) [inline src/x/xHilbertTransform.ts:18]: Should be the same as:

`X[i * 2] *= v[i];`

File: src/x/xHilbertTransform.ts [hunk 2]
```diff
@@ -1,28 +1,32 @@
...
-      aSum += array[i] * log + (array[i + 1] - array[i]) * (-1 + (k - i) * log);
-    }
-    const b = array[k - 1] - array[k + 1];
-    let cSum = 0;
-    for (let i = k + 1; i < array.length - 1; i++) {
-      const log = Math.log((i - k) / (i - k + 1));
-      cSum += array[i] * log + (array[i - 1] - array[i]) * (1 + (i - k) * log);
-    }
-    result[k - 1] = ((inClockwise ? 1 : -1) * (aSum + b + cSum)) / Math.PI;
+export function xHilbertTransform(x: number[]) {
```
Comment (lpatiny) [inline src/x/xHilbertTransform.ts:8]: How to deal with non power of 2 ?

File: src/x/xHilbertTransform.ts [hunk 3]
```diff
@@ -1,28 +1,32 @@
...
+  if (!Array.isArray(x)) {
+    throw new Error('x must be a vector');
+  }
+  const n = x.length;
+  const fft = new FFT(n);
+  const v = [1, ...Array(n / 2 - 1).fill(2), 1, ...Array(n / 2 - 1).fill(0)];
```
Comment (lpatiny) [inline src/x/xHilbertTransform.ts:14]: You could use directly a typed array
```
const v = new Float64Array(n);
v[0]=1
v[n/2]=1
for (let i=1; i<n/2; i++) {
v[i]=2;
}
```

File: src/x/__tests__/xHilbertTransform.test.ts [hunk 4]
```diff
@@ -1,34 +1,32 @@
 import { xHilbertTransform } from '../../index';
 
-describe('test hilbert transform of cos(t)', () => {
-  const length = 50;
+describe('test hilbert transform', () => {
```
Comment (lpatiny) [inline src/x/__tests__/xHilbertTransform.test.ts:3]: Keep the old test case as well. If the old testcase was correct and there is no way of have the FFT approach with non power of 2 there should be a new method xFastHilbertTransform that could only be applied on power of 2 data size.

**Review by lpatiny — CHANGES_REQUESTED** [review]

**Comment by targos** [comment]
You need to explain the breaking change in the commit message (see the conventional commit spec) and not use "!" unless the title is already clear about what's breaking

**Comment by josoriom** [comment]
> You need to explain the breaking change in the commit message (see the conventional commit spec) and not use "!" unless the title is already clear about what's breaking

Thanks, I'll remember that!

File: src/x/xHilbertTransform.ts [hunk 5]
```diff
@@ -1,26 +1,66 @@
+import { DoubleArray } from 'cheminfo-types';
+import FFT from 'fft.js';
+
+import { xCheck } from './xCheck';
+
 /**
  * Performs the Hilbert transform
  * @returns A new vector with 90 degree shift regarding the phase of the original function
```
Comment (lpatiny) [inline src/x/xHilbertTransform.ts:8]: We should take the habit to add reference if we can: 

 * @link https://en.wikipedia.org/wiki/Hilbert_transform

File: src/x/xHilbertTransform.ts [hunk 6]
```diff
@@ -1,26 +1,66 @@
...
  */
 
-export function xHilbertTransform(
-  input: number[],
+export function xHilbertTransform(array: DoubleArray) {
+  xCheck(array);
+  if (Math.log2(array.length) % 1 === 0) {
+    return hilbertTransformWithFFT(array);
+  } else {
+    return hilbertTransform(array);
```
Comment (lpatiny) [inline src/x/xHilbertTransform.ts:16]: You didn't find a way to make a zero filling to next power of 2, make the hilbertTransformWithFFT and then cut the result ?
@jobo322 Could you give us your opinion about the feasability ?

**Comment by lpatiny** [comment]
There is no breaking change and speed improvement so it is great to me.  However I'm still wondering if we could do a trick that allows to always use FFT.

File: src/x/xHilbertTransform.ts (hunk 6 above)
Comment (jobo322) [reply src/x/xHilbertTransform.ts:16]: we have two functions not exposed in nmr-processing

```
function nextPowerOfTwo(n: number) {
  if (n === 0) return 1;
  n--;
  n |= n >> 1;
  n |= n >> 2;
  n |= n >> 4;
  n |= n >> 8;
  n |= n >> 16;
  return n + 1;
}
```

```
function isPowerOfTwo(n: number) {
  return n !== 0 && (n & (n - 1)) === 0;
}
```

we could move those functions to spectra-processing and use it in combination with the [xreimZeroFilling](https://github.com/mljs/spectra-processing/blob/main/src/xreim/xreimZeroFilling.ts) function

Comment (jobo322) [reply src/x/xHilbertTransform.ts:16]: xreimZeroFilling is redundant in this case

**Pull request merged**